sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -650,15 +646,15 @@ object FoldablePropagation extends Rule[LogicalPlan] {
 
     if (foldableExprSet.isEmpty) {
       plan
-    } else if (hasCommandOrUnion(plan)) {
-      replaceInSubplan(plan)
     } else {
       val foldableMap = AttributeMap(foldableExprSet.toSeq.map(_.asInstanceOf[Alias])
         .map(a => (a.toAttribute, a.child)))
 
-      CleanupAliases(plan.transformAllExpressions {
-        case a : AttributeReference if foldableMap.contains(a) =>
-          Alias(foldableMap.getOrElse(a.toAttribute, a), a.name)()
+      CleanupAliases(plan.transformUp {
+        case p: LogicalPlan if !hasCommandOrUnion(p) => p.transformAllExpressions {
+          case a: AttributeReference if foldableMap.contains(a) =>
+            Alias(foldableMap.getOrElse(a.toAttribute, a), a.name)()
+        }
```

Comment:
I'd like to write:
```
var stop = false
plan.transformUp {
  case u: Union =>
    stop = true
    u
  case c: Command =>
    stop = true
    c
  case p if !stop => p.transformExpressions {
    case a: AttributeReference if foldableMap.contains(a) =>
      foldableMap(a) // foldableMap's value shoud be Alias, not its child
  }
}
```

if earlier sibling is `Union`, this is suboptimal, but still safe(it's ok to not apply this optimization for some corner cases)